### PR TITLE
[SYCL] Implement fallback path for profiling info support

### PR DESCRIPTION
**Pull request by eunakim0103**
OpenCL versions less than 2.1 do not have the API clGetDeviceAndHostTimer, therefore we provided limited profiling which does not support info::event_profiling::command_submit.  If this time was requested, an exception was thrown.  With this fix, we now support a complete fallback for profiling info, including info::event_profiling::command_submit.

The full fallback solution for profiling information has the following steps (_only applies when OpenCL version < 2.1_):

1. Set Host_Submit_Timestamp: Capture a host timestamp (no backend involvement at all) in the SYCL runtime as part of queue::submit.  This determines the timebase for event profiling.  This is what gets returned for info::event_profiling::command_submit.
2. Set Host_Queue_Timestamp: Capture another host timestamp (no backend involvement here either) in the SYCL runtime when the command actually gets queued to the backend.  This will never be returned directly, but it will be used to normalize to the event profiling timebase.
3. Set Device_Queue_Time and Device_Start_Time: For the return value for info::event_profiling::command_start, use the backend event profiling START information, normalized by the backend event profiling QUEUED information and the timestamp captured in (2).
4. Set Device_Queue_Time and Device_End_Time: For the return value for info::event_profiling::command_end, same as (3), except use the backend event profiling END information instead.

The expected return values are the following:
1. <info::event_profiling::command_submit> returns Host_Submit_Timestamp (step 1)
2. <info::event_profiling::command_start> returns (Host_Queue_Timestamp – Device_Queue_Time) + Device_Start_Time
3. <info::event_profiling::command_end> returns (Host_Queue_Timestamp – Device_Queue_Time) + Device_End_Time

### CODE REVIEW

**Review by sergey-semenov — COMMENTED** [review]
Did a brief review of the current state, ignoring the debugging code and disabled tests. I wouldn't recommend disabling them like that in the future, even for draft PRs, since it makes it difficult to check their status once the PR is updated.

File: sycl/source/detail/queue_impl.hpp
Comment (eunakim0103) [reply sycl/source/detail/queue_impl.hpp:156]: Good point! Will fix

File: sycl/source/detail/device_info.hpp
Comment (eunakim0103) [reply sycl/source/detail/device_info.hpp:243]: I can do that but there is more than 1 places that calls  checkNativeQueueProfiling (queue_impl.hpp and device_info.hpp)

File: sycl/source/detail/event_impl.cpp
Comment (eunakim0103) [reply sycl/source/detail/event_impl.cpp]: all steps have been added

File: sycl/source/detail/device_impl.cpp
Comment (dm-vodopyanov) [inline sycl/source/detail/device_impl.cpp:636]:
```suggestion
  return Result != PI_ERROR_INVALID_OPERATION;
```

File: sycl/source/detail/device_impl.hpp
Comment (dm-vodopyanov) [inline sycl/source/detail/device_impl.hpp:244]:
```suggestion

  bool IsGetDeviceAndHostTimerSupported();
```

File: sycl/source/handler.cpp
Comment (eunakim0103) [inline sycl/source/handler.cpp:291]: The existing submit time was captured after enqueue (kernel launching). The host side sycl timestamp second timestamp is captured earliest possible without changing the original submit time.

File: sycl/source/detail/event_impl.cpp
Comment (dm-vodopyanov) [inline sycl/source/detail/event_impl.cpp:104]:
```suggestion
  auto Timestamp = std::chrono::high_resolution_clock::now().time_since_epoch();
```

File: sycl/source/detail/event_impl.cpp
Comment (dm-vodopyanov) [inline sycl/source/detail/event_impl.cpp:471]:
```suggestion
    // that API clGetDeviceAndHostTimer is not available to obtain the precise
```

File: sycl/source/detail/event_impl.cpp
Comment (dm-vodopyanov) [inline sycl/source/detail/event_impl.cpp:472]:
```suggestion
    // host and device submit time. The host timestamp in SYCL runtime will be
```

File: sycl/source/detail/event_impl.cpp
Comment (dm-vodopyanov) [inline sycl/source/detail/event_impl.cpp:491]:
```suggestion
  // that API clGetDeviceAndHostTimer is not available to obtain the precise
```

File: sycl/source/detail/event_impl.cpp
Comment (dm-vodopyanov) [inline sycl/source/detail/event_impl.cpp:492]:
```suggestion
  // host and device submit time. The host timestamp in SYCL runtime will be
```

File: sycl/source/detail/event_impl.cpp
Comment (dm-vodopyanov) [inline sycl/source/detail/event_impl.cpp:481]: Please add dots in the end of the sentences and start new sentences with with a capital letter

File: sycl/source/detail/event_impl.hpp
Comment (dm-vodopyanov) [inline sycl/source/detail/event_impl.hpp:231]:
```suggestion
  /// Calling this function to capture the host timestamp to use
```

File: sycl/source/detail/event_impl.hpp
Comment (dm-vodopyanov) [inline sycl/source/detail/event_impl.hpp:232]:
```suggestion
  /// profiling base time. See MFallbackProfiling
```

File: sycl/source/detail/event_impl.hpp
Comment (dm-vodopyanov) [inline sycl/source/detail/event_impl.hpp:303]: Empty line looks not needed

File: sycl/source/detail/queue_impl.hpp
Comment (dm-vodopyanov) [inline sycl/source/detail/queue_impl.hpp:158]: Such comments with many implementation details can be easily outdated.

File: sycl/test-e2e/Regression/dummy_event_info.cpp
Comment (dm-vodopyanov) [inline sycl/test-e2e/Regression/dummy_event_info.cpp:27]: ?

File: sycl/unittests/queue/GetProfilingInfo.cpp
Comment (dm-vodopyanov) [inline sycl/unittests/queue/GetProfilingInfo.cpp:26]: ?

File: sycl/unittests/queue/GetProfilingInfo.cpp
Comment (dm-vodopyanov) [inline sycl/unittests/queue/GetProfilingInfo.cpp:317]: ?

And for all other commented code below

File: sycl/unittests/queue/GetProfilingInfo.cpp
Comment (dm-vodopyanov) [inline sycl/unittests/queue/GetProfilingInfo.cpp:445]:
```suggestion
  static sycl::unittest::PiImage DevImage =
```
There is no `DevImage_2` in the code

File: sycl/unittests/queue/GetProfilingInfo.cpp
Comment (dm-vodopyanov) [inline sycl/unittests/queue/GetProfilingInfo.cpp:448]: The same for this variable

File: sycl/unittests/queue/GetProfilingInfo.cpp
Comment (dm-vodopyanov) [inline sycl/unittests/queue/GetProfilingInfo.cpp:455]: Variable is not used

File: sycl/unittests/queue/GetProfilingInfo.cpp
Comment (dm-vodopyanov) [inline sycl/unittests/queue/GetProfilingInfo.cpp:430]:
```suggestion
TEST(GetProfilingInfo, fallback_profiling_PiGetDeviceAndHostTimer_unsupported) {
```

File: sycl/unittests/queue/GetProfilingInfo.cpp
Comment (dm-vodopyanov) [inline sycl/unittests/queue/GetProfilingInfo.cpp:428]: What does this test checks? Can we check the validness of submit_time, start_time and end_time?

Comment (eunakim0103) [reply sycl/unittests/queue/GetProfilingInfo.cpp:428]: Yes, we can do that. Let me add

File: sycl/source/handler.cpp
Comment (sergey-semenov) [reply sycl/source/handler.cpp:291]: Doesn't this make the queue base time essentially the same as submission time?
Submission time here captures the moment when the command is submitted to the SYCL runtime. The queue base time should reflect the point at which the command is enqueued to the underlying backend, i.e. right before we make the corresponding piEnqueue* call.

**Review by sergey-semenov — CHANGES_REQUESTED** [review]

File: sycl/source/handler.cpp
Comment (eunakim0103) [reply sycl/source/handler.cpp:291]: Unless we move the NewEvent->setSubmissionTime(); position which captures host timestamp 1 in step 1, capturing host base timestamp before piEnqueue will be earlier time than submit time since submit time is captured after kernel launching. Still it would be a good idea to do that? Also it was a bit tricky to add the timestamp capturing code in handler due to the mutex. Once  in a while the event could not be completed.

**Comment by dm-vodopyanov** [comment]
@eunakim0103 please, close all resolved threads in this PR and in all upcoming. Pros: PR becomes not messy, gatekeepers get an ability to easily track that all comments from reviewers were resolved before merging.

File: sycl/source/detail/scheduler/commands.cpp
Comment (sergey-semenov) [inline sycl/source/detail/scheduler/commands.cpp:1320]: In scheduler, the only command type that is visible to the user via an event is ExecCGCommand. Others are internal, so there's no need to handle those (like MapMemObject, UnMapMemObject, MemCpyCommand, MemCpyCommandHost).

All of the CG cases in the `ExecCGCommand::enqueueImpQueue` do need to be handled though.

**Review by reble — APPROVED** [review]
LGTM for SYCL Graph related changes

**Review by againull — APPROVED** [review]

File: sycl/source/detail/scheduler/commands.cpp
Comment (againull) [reply sycl/source/detail/scheduler/commands.cpp]: Addressed.

File: sycl/source/detail/context_impl.cpp
Comment (dm-vodopyanov) [inline sycl/source/detail/context_impl.cpp:376]: No need to remove this line

File: sycl/source/detail/context_impl.cpp
Comment (dm-vodopyanov) [inline sycl/source/detail/context_impl.cpp:388]: No need to remove this line

File: sycl/source/detail/device_impl.hpp
Comment (dm-vodopyanov) [inline sycl/source/detail/device_impl.hpp:246]: @eunakim0103 we discussed with you offline a couple of days ago how it should be properly addressed, still, didn't fixed.

File: sycl/source/detail/device_info.hpp
Comment (dm-vodopyanov) [inline sycl/source/detail/device_info.hpp:233]: This wasn't fixed at all, but the conversation was resolved, @eunakim0103 
https://github.com/intel/llvm/pull/10662#discussion_r1291180455

File: sycl/source/detail/event_impl.cpp
Comment (dm-vodopyanov) [inline sycl/source/detail/event_impl.cpp:296]: Inconsistent variable naming in the single block of code

File: sycl/source/detail/event_impl.cpp
Comment (dm-vodopyanov) [inline sycl/source/detail/event_impl.cpp:323]: the same

File: sycl/test/abi/sycl_symbols_linux.dump
Comment (dm-vodopyanov) [inline sycl/test/abi/sycl_symbols_linux.dump:4721]: Why it is removed? Is it ABI break between the releases?

File: sycl/test/abi/sycl_symbols_windows.dump
Comment (dm-vodopyanov) [inline sycl/test/abi/sycl_symbols_windows.dump:1028]: Why it is removed? Is it ABI break between the releases?

**Review by dm-vodopyanov — COMMENTED** [review]
@againull why it was merged without @sergey-semenov and mine approvals? Let's not do this like that again?

File: sycl/test/abi/sycl_symbols_linux.dump
Comment (againull) [reply sycl/test/abi/sycl_symbols_linux.dump:4721]: This symbol is moved to another line because  the script was used to generate this file.

File: sycl/test/abi/sycl_symbols_windows.dump
Comment (againull) [reply sycl/test/abi/sycl_symbols_windows.dump:1028]: This symbol is moved to another line because  the script was used to generate this file.

File: sycl/source/detail/event_impl.cpp
Comment (againull) [reply sycl/source/detail/event_impl.cpp:296]: Fixed in follow up patch

File: sycl/unittests/queue/GetProfilingInfo.cpp
Comment (sergey-semenov) [inline sycl/unittests/queue/GetProfilingInfo.cpp:399]: This appears to be unused.

File: sycl/unittests/queue/GetProfilingInfo.cpp
Comment (sergey-semenov) [inline sycl/unittests/queue/GetProfilingInfo.cpp:1]: The added test only checks kernel launch, but since the time capture needed to be added to all sorts of different places (other command group types, queue shortcut functions), we also need test cases that verify those paths.

**Pull request merged**
